src/components/nano/NanoContractHistory.js

```diff
@@ -0,0 +1,238 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+import Loading from '../../components/Loading';
+import { Link, useLocation } from 'react-router-dom';
+import { NANO_CONTRACT_TX_HISTORY_COUNT } from '../../constants';
+import TxRow from '../tx/TxRow';
+import helpers from '../../utils/helpers';
+import nanoApi from '../../api/nanoApi';
+import WebSocketHandler from '../../WebSocketHandler';
+import PaginationURL from '../../utils/pagination';
+import hathorLib from '@hathor/wallet-lib';
+import { reverse } from 'lodash';
+
+
+/**
+ * Displays nano tx history in a table with pagination buttons. As the user navigates through the history,
+ * the URL parameters 'hash' and 'page' are updated.
+ *
+ * Either all URL parameters are set or they are all missing.
+ *
+ * Example 1:
+ *   hash = "00000000001b328fafb336b4515bb9557733fe93cf685dfd0c77cae3131f3fff"
+ *   page = "previous"
+ *
+ * Example 2:
+ *   hash = "00000000001b328fafb336b4515bb9557733fe93cf685dfd0c77cae3131f3fff"
+ *   page = "next"
+ */
+function NanoContractHistory({ ncId }) {
+  // We must use memo here because we were creating a new pagination
+  // object in every new render, so the useEffect was being called forever
+  const pagination = useMemo(() =>
+    new PaginationURL({
+      'hash': {'required': false},
+      'page': {'required': false}
+    })
+  , []);
+
+  const location = useLocation();
+
+  // loading {boolean} Bool to show/hide loading element
+  const [loading, setLoading] = useState(true);
+  // history {Array} Nano contract history
+  const [history, setHistory] = useState([]);
+  // errorMessage {string} Message to show when error happens on history load
+  const [errorMessage, setErrorMessage] = useState('');
+  // hasBefore {boolean} If 'Previous' button should be enabled 
+  const [hasBefore, setHasBefore] = useState(false);
+  // hasAfter {boolean} If 'Next' button should be enabled 
+  const [hasAfter, setHasAfter] = useState(false);
+
+  /**
+   * useCallback is important here to update this method with new history state
+   * otherwise it would be fixed the moment the event listener is started in the useEffect
+   * with the history as an empty array
+   *
+   * tx {Transaction} Transaction object that arrived from the websocket
+   */
+  const updateListWs = useCallback((tx) => {
+    // We only add to the list if it's the first page and it's a new tx from this nano
+    if (!hasBefore) {
```

Comment:
suggestion(non-blocking): avoid nested code blocks for simple conditions by early exiting the function
```js
if (hasBefore) {
  return;
}

if (tx.version !== hathorLib.constants.NANO_CONTRACTS_VERSION || tx.nc_id !== ncId) {
  return;
}

// ... actual code and logic here
```

Reply:
Much better, thanks https://github.com/HathorNetwork/hathor-explorer/pull/292/commits/993898f82bd11ba930d99bf9458493c078fbffc7